In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('../data/spam-train.csv')

In [3]:
df.head()

input target
0                      Ok lar... Joking wif u oni...    ham
1  Free entry in 2 a wkly comp to win FA Cup fina...   spam
2  U dun say so early hor... U c already then say...    ham
3  Nah I don't think he goes to usf, he lives aro...    ham
4  FreeMsg Hey there darling it's been 3 week's n...   spam

In [4]:
from sklearn.feature_extraction.text import CountVectorizer

In [5]:
vec = CountVectorizer(stop_words='english')

In [6]:
X = vec.fit_transform(df['input'])

In [7]:
X

<4458x7505 sparse matrix of type '<class 'numpy.int64'>'
	with 34923 stored elements in Compressed Sparse Row format>

In [8]:
df.shape

(4458, 2)

In [9]:
len(vec.vocabulary_)

7505

In [10]:
vec.vocabulary_['ok']

4774

In [11]:
ex0 = X[0].todense()

In [12]:
import numpy as np

In [13]:
np.array(ex0)[0][4774]

1

In [14]:
from sklearn.preprocessing import LabelEncoder

In [15]:
le = LabelEncoder()
y = le.fit_transform(df['target'])

In [16]:
y

array([0, 1, 0, ..., 0, 0, 0])

In [17]:
from sklearn.model_selection import train_test_split

In [18]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=123)

In [19]:
from sklearn.neighbors import KNeighborsClassifier

In [20]:
knn = KNeighborsClassifier(metric='cosine')

In [21]:
knn.fit(X_train, y_train)

KNeighborsClassifier(metric='cosine')

In [22]:
knn.score(X_test, y_test)

0.9697309417040358

In [23]:
y_pred = knn.predict(X_test)

In [24]:
from sklearn.metrics import confusion_matrix

In [25]:
confusion_matrix(y_test, y_pred)

array([[778,   2],
       [ 25,  87]])

In [26]:
y_test.shape

(892,)

**Interpretation:** 
- Class 0 ('ham') was predicted as 0 in 778 examples.
- Class 1 ('spam') was predicted as 1 in 87 examples.
- Two values of class 'ham' classified as 'spam'.
- 25 values of class 'spam' classified as 'ham'.

In [27]:
27/892

0.030269058295964126

In [37]:
msg = 'get best offers call 123456789'

In [38]:
x_msg = vec.transform([msg])

In [39]:
x_msg

<1x7505 sparse matrix of type '<class 'numpy.int64'>'
	with 2 stored elements in Compressed Sparse Row format>

In [40]:
knn.predict(x_msg)

array([0])